File: src/core/vdom/bounds.ts
export type LayoutDirection = 'horizontal' | 'vertical' | 'overlap'

export type Measurement =
  Measurement2 |
  `${Measurement2} ${'+' | '-'} ${Measurement3}` |
  `${Measurement2} ${'+' | '-'} ${Measurement3} ${'+' | '-'} ${Measurement4}` |
  `${Measurement2} ${'+' | '-'} ${Measurement3} ${'+' | '-'} ${Measurement4} ${'+' | '-'} ${Measurement5}`

type Measurement2 = 'prev' | Measurement3
type Measurement3 = `${number}%` | Measurement4
type Measurement4 = `${number}px` | Measurement5
type Measurement5 = `${number}` | number

export type LayoutPosition1D =
  'global-absolute' |
  'local-absolute' |
  'relative'

export type LayoutPosition =
  LayoutPosition1D |
  { x: LayoutPosition1D, y: LayoutPosition1D }

export interface SubLayout {
  direction?: LayoutDirection
  gap?: Measurement
  custom?: any
}

export interface BoundingBox {
  x: number
  y: number
  z: number
  anchorX: number
  anchorY: number
  width?: number
  height?: number
}

export interface Size {
  width: number
  height: number
}

export interface Rectangle {
  left: number
  top: number
  width: number
  height: number
}

export interface ParentBounds {
  boundingBox: BoundingBox
  sublayout: SubLayout
  columnSize: Size
}

export type Bounds = (parent: ParentBounds, prevSibling: Rectangle | null) => BoundingBox

export interface FullBoundsSpec {
  layout?: LayoutPosition
  x?: Measurement
  y?: Measurement
  z?: number
  anchorX?: number
  anchorY?: number
  width?: Measurement
  height?: Measurement
}

export type BoundsSpec = FullBoundsSpec

export function Bounds (spec: BoundsSpec): Bounds {
  return (parent, prevSibling) => ({
    x: applyLayoutX(parent, prevSibling, spec.layout, reifyX(parent, 'not-applicable', spec.x)),
    y: applyLayoutY(parent, prevSibling, spec.layout, reifyY(parent, 'not-applicable', spec.y)),
    z: spec.z ?? parent.boundingBox.z + Bounds.BOX_Z,
    anchorX: spec.anchorX ?? 0,
    anchorY: spec.anchorY ?? 0,
    width: spec.width === undefined ? undefined : reifyX(parent, prevSibling?.width ?? null, spec.width),
    height: spec.height === undefined ? undefined : reifyY(parent, prevSibling?.height ?? null, spec.height)
  })
}

function reifyX (parent: ParentBounds, prevSibling: number |'not-applicable' | null, x: Measurement | undefined): number {
  if (x === undefined) {
    return 0
  } else if (typeof x === 'number') {
    return x
  } else if (/^[0-9]+$/.test(x)) {
    return parseInt(x)
  } else if (/^[0-9]*\.[0-9]+$/.test(x)) {
    return parseFloat(x)
  } else if (x.endsWith('%')) {
    if (parent.boundingBox.width === undefined) {
      throw new Error(`cannot convert percent ${x} to number because parent width is unknown`)
    }
    return (parent.boundingBox.width * parseFloat(x) / 100)
  } else if (x.endsWith('px')) {
    return parseFloat(x) / parent.columnSize.width
  } else if (x === 'prev') {
    if (prevSibling === 'not-applicable') {
      throw new Error('can\'t use \'prev\' for position or gap')
    } else if (prevSibling === null) {
      throw new Error('can\'t use \'prev\' for first child')
    } else {
      return prevSibling
    }
  } else if (x.includes('+')) {
    const [left, right] = x.split('+')
    return reifyX(parent, prevSibling, left.trimEnd() as Measurement) + reifyX(parent, prevSibling, right.trimStart() as Measurement)
  } else if (x.includes('-')) {
    const [left, right] = x.split('-')
    return reifyX(parent, prevSibling, left.trimEnd() as Measurement) - reifyX(parent, prevSibling, right.trimStart() as Measurement)
  } else {
    throw new Error(`invalid measurement: ${x}`)
  }
}

function reifyY (parent: ParentBounds, prevSibling: number | 'not-applicable' | null, y: Measurement | undefined): number {
  if (y === undefined) {
    return 0
  } else if (typeof y === 'number') {
    return y
  } else if (/^[0-9]+$/.test(y)) {
    return parseInt(y)
  } else if (/^[0-9]*\.[0-9]+$/.test(y)) {
    return parseFloat(y)
  } else if (y.endsWith('%')) {
    if (parent.boundingBox.height === undefined) {
      throw new Error(`bad layout: cannot convert percent ${y} to number because parent height is unknown`)
    }
    return (parent.boundingBox.height * parseFloat(y) / 100)
  } else if (y.endsWith('px')) {
    return parseFloat(y) / parent.columnSize.height
  } else if (y === 'prev') {
    if (prevSibling === 'not-applicable') {
      throw new Error('can\'t use \'prev\' for position or gap')
    } else if (prevSibling === null) {
      throw new Error('can\'t use \'prev\' for first child')
    } else {
      return prevSibling
    }
  } else if (y.includes('+')) {
    const [left, right] = y.split('+')
    return reifyY(parent, prevSibling, left.trimEnd() as Measurement) + reifyY(parent, prevSibling, right.trimStart() as Measurement)
  } else if (y.includes('-')) {
    const [left, right] = y.split('-')
    return reifyY(parent, prevSibling, left.trimEnd() as Measurement) - reifyY(parent, prevSibling, right.trimStart() as Measurement)
  } else {
    throw new Error(`invalid measurement: ${y}`)
  }
}

function applyLayoutX (parent: ParentBounds, prevSibling: Rectangle | null, layout: LayoutPosition | undefined, reified: number): number {
  const layout1D = typeof layout === 'string' || typeof layout === 'undefined' ? layout : layout.x
  switch (layout1D) {
    case 'global-absolute':
      return reified
    case 'local-absolute':
      return reified + parent.boundingBox.x
    case undefined:
    case 'relative':
      switch (parent.sublayout.direction) {
        case 'horizontal': {
          // Yes, we do want to reify the parent's sublayout with it's own bounds
          const gap = reifyX(parent, null, parent.sublayout.gap)
          return reified + (prevSibling !== null ? prevSibling.left + prevSibling.width + gap : getLayoutBoundingBoxLeft(parent.boundingBox))
        }
        case 'vertical':
          return reified + parent.boundingBox.x
        case 'overlap':
          return reified + parent.boundingBox.x
        case undefined:
          if (prevSibling !== null) {
            console.warn('no layout direction specified with multiple children, defaulting to overlap (applyLayoutX)')
          }
          return reified + parent.boundingBox.x
      }
  }
}

function applyLayoutY (parent: ParentBounds, prevSibling: Rectangle | null, layout: LayoutPosition | undefined, reified: number): number {
  const layout1D = typeof layout === 'string' || typeof layout === 'undefined' ? layout : layout.y
  switch (layout1D) {
    case 'global-absolute':
      return reified
    case 'local-absolute':
      return reified + parent.boundingBox.y
    case undefined:
    case 'relative':
      switch (parent.sublayout.direction) {
        case 'horizontal':
          return reified + parent.boundingBox.y
        case 'vertical': {
          // Yes, we do want to reify the parent's sublayout with it's own bounds
          const gap = reifyY(parent, null, parent.sublayout.gap)
          return reified + (prevSibling !== null ? prevSibling.top + prevSibling.height + gap : getLayoutBoundingBoxTop(parent.boundingBox))
        }
        case 'overlap':
          return reified + parent.boundingBox.y
        case undefined:
          if (prevSibling !== null) {
            console.warn('no layout direction specified with multiple children, defaulting to overlap (applyLayoutY)')
          }
          return reified + parent.boundingBox.y
      }
  }
}

export function getLayoutBoundingBoxLeft (bounds: BoundingBox): number {
  if (bounds.anchorX === 0) {
    return bounds.x
  } else if (bounds.width === undefined) {
    throw new Error('bad layout: bounds not anchored at left has no width, so we don\'t know where to put the child')
  } else {
    return bounds.x - (bounds.anchorX * bounds.width)
  }
}

export function getLayoutBoundingBoxTop (bounds: BoundingBox): number {
  if (bounds.anchorY === 0) {
    return bounds.y
  } else if (bounds.height === undefined) {
    throw new Error('bad layout: bounds not anchored at top has no height, so we don\'t know where to put the child')
  } else {
    return bounds.y - (bounds.anchorY * bounds.height)
  }
}

export module ParentBounds {
  export function equals (a: ParentBounds, b: ParentBounds): boolean {
    return JSON.stringify(a) === JSON.stringify(b)
  }
}

export module Rectangle {
  export function equals (a: Rectangle | null, b: Rectangle | null): boolean {
    return JSON.stringify(a) === JSON.stringify(b)
  }

  export function union (a: Rectangle | null, b: Rectangle | null): Rectangle | null {
    if (a === null) {
      return b
    } else if (b === null) {
      return a
    } else {
      const left = Math.min(a.left, b.left)
      const top = Math.min(a.top, b.top)
      const right = Math.max(a.left + a.width, b.left + b.width)
      const bottom = Math.max(a.top + a.height, b.top + b.height)
      return {
        left,
        top,
        width: right - left,
        height: bottom - top
      }
    }
  }
}

export module BoundingBox {
  export function toRectangle (bounds: BoundingBox & Size): Rectangle
  export function toRectangle (bounds: BoundingBox, size: Size): Rectangle
  export function toRectangle (bounds: BoundingBox, size?: Size): Rectangle {
    const width = bounds.width ?? size?.width
    const height = bounds.height ?? size?.height
    if (width === undefined || height === undefined) {
      throw new Error('toRectangle invalid arguments: bounds has no size and no size provided')
    }
    return {
      left: bounds.x - (bounds.anchorX * width),
      top: bounds.y - (bounds.anchorY * height),
      width,
      height
    }
  }
}

export module Bounds {
  export const BOX_Z = 0.0001
  export const DELTA_Z = 0.0000001

  export const DEFAULT: Bounds = parent => ({
    x: parent.boundingBox.x,
    y: parent.boundingBox.y,
    z: parent.boundingBox.z + BOX_Z,
    anchorX: parent.boundingBox.anchorX,
    anchorY: parent.boundingBox.anchorY
  })

  export const FILL_X: Bounds = parent => ({
    x: parent.boundingBox.x,
    y: parent.boundingBox.y,
    z: parent.boundingBox.z + BOX_Z,
    anchorX: parent.boundingBox.anchorX,
    anchorY: parent.boundingBox.anchorY,
    width: parent.boundingBox.width
  })

  export const FILL_Y: Bounds = parent => ({
    x: parent.boundingBox.x,
    y: parent.boundingBox.y,
    z: parent.boundingBox.z + BOX_Z,
    anchorX: parent.boundingBox.anchorX,
    anchorY: parent.boundingBox.anchorY,
    height: parent.boundingBox.height
  })

  export const FILL: Bounds = parent => ({
    x: parent.boundingBox.x,
    y: parent.boundingBox.y,
    z: parent.boundingBox.z + BOX_Z,
    anchorX: parent.boundingBox.anchorX,
    anchorY: parent.boundingBox.anchorY,
    width: parent.boundingBox.width,
    height: parent.boundingBox.height
  })
}

File: devolve-ui/src/core/view/layout/bounds.rs
use std::collections::HashMap;
use crate::core::view::layout::err::{LayoutError, LayoutResult};
use crate::core::view::layout::geom::{BoundingBox, Rectangle};
use crate::core::view::layout::measurement::{Measurement, MeasurementUnit};
use crate::core::view::layout::parent_bounds::{DimsStore, LayoutDirection, ParentBounds};

#[derive(Debug, Clone, PartialEq)]
pub struct Bounds {
    pub layout: LayoutPosition,
    pub x: Measurement,
    pub y: Measurement,
    /// By default, the nodes are rendered next after prev, child after parent but before parent's sibling.
    /// Actually-specified z position takes precedence over this. If 2 nodes have the same z-position,
    /// they will be rendered as specified by the above.
    pub z: i32,
    pub anchor_x: f32,
    pub anchor_y: f32,
    pub width: Option<Measurement>,
    pub height: Option<Measurement>
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub enum LayoutPosition1D {
    Relative,
    GlobalAbsolute,
    LocalAbsolute
}

#[derive(Debug, Clone, Default, Eq, PartialEq)]
pub struct LayoutPosition {
    pub x: LayoutPosition1D,
    pub y: LayoutPosition1D,
}

const MAX_CHILDREN_EXPECTED_LOG2: f64 = 8f64;

#[derive(Debug, Clone, Copy)]
enum PrevSiblingDim {
    NotApplicable,
    FirstChild,
    Some(f32)
}

impl From<Option<f32>> for PrevSiblingDim {
    fn from(prev_sibling_dim: Option<f32>) -> Self {
        match prev_sibling_dim {
            None => PrevSiblingDim::FirstChild,
            Some(prev_sibling_dim) => PrevSiblingDim::Some(prev_sibling_dim),
        }
    }
}

impl Bounds {
    pub fn resolve(&self, parent_bounds: &ParentBounds, prev_sibling: Option<&Rectangle>, parent_depth: usize, sibling_index: usize) -> LayoutResult<(BoundingBox, DimsStore)> {
        let mut store = parent_bounds.store.clone();
        let bounding_box = BoundingBox {
            x: Self::apply_layout_x(parent_bounds, prev_sibling, self.layout.x, Self::reify_x(parent_bounds, &PrevSiblingDim::NotApplicable, Some(&mut store.x), &self.x).map_err(|err| err.add_store("x"))?).map_err(|err| err.add_store("x@layout"))?,
            y: Self::apply_layout_y(parent_bounds, prev_sibling, self.layout.y, Self::reify_y(parent_bounds, &PrevSiblingDim::NotApplicable, Some(&mut store.y), &self.y).map_err(|err| err.add_store("y"))?).map_err(|err| err.add_store("y@layout"))?,
            z: self.z as f64 + parent_bounds.bounding_box.z - f64::floor(parent_bounds.bounding_box.z) + (((parent_depth + 1) as f64 * -MAX_CHILDREN_EXPECTED_LOG2).exp2() * (sibling_index + 1) as f64),
            anchor_x: self.anchor_x,
            anchor_y: self.anchor_y,
            width: self.width.as_ref().map(|width| Self::reify_x(parent_bounds, &prev_sibling.map(|r| r.width()).into(), Some(&mut store.width), &width).map_err(|err| err.add_dimension("width"))).transpose()?.into(),
            height: self.height.as_ref().map(|height| Self::reify_y(parent_bounds, &prev_sibling.map(|r| r.height()).into(), Some(&mut store.height), &height).map_err(|err| err.add_dimension("height"))).transpose()?.into()
        };
        Ok((bounding_box, store))
    }

    fn reify_x(parent_bounds: &ParentBounds, prev_sibling: &PrevSiblingDim, store: Option<&mut HashMap<&'static str, f32>>, x: &Measurement) -> LayoutResult<f32> {
        let mut reified = 0f32;
        for x in x.iter_adds() {
            reified += x.value.scalar() * match x.unit {
                MeasurementUnit::Units => 1f32,
                MeasurementUnit::Pixels => 1f32 / parent_bounds.column_size.width,
                MeasurementUnit::PercentOfParent => match parent_bounds.bounding_box.width.into_option() {
                    None => Err(LayoutError::new("can't use fraction for x: parent width not known"))?,
                    Some(width) => width / 100f32
                },
                MeasurementUnit::OfPrev => match prev_sibling {
                    PrevSiblingDim::NotApplicable => Err(LayoutError::new("can't use prev for x: not applicable"))?,
                    PrevSiblingDim::FirstChild => 0f32,
                    PrevSiblingDim::Some(prev_sibling_dim) => *prev_sibling_dim
                },
                MeasurementUnit::OfLoad(ident) => match &store {
                    None => Err(LayoutError::new("can't load y: dim-store not applicable"))?,
                    Some(store) => match store.get(ident) {
                        None => Err(LayoutError::new(format!("can't load y: no such dim {}", ident)))?,
                        Some(result) => *result
                    }
                }
            };
        }
        if let Some(ident) = x.store {
            match store {
                None => Err(LayoutError::new("can't store x: dim-store not applicable"))?,
                Some(store) => store.insert(ident, reified)
            };
        }
        Ok(reified)
    }

    fn reify_y(parent_bounds: &ParentBounds, prev_sibling: &PrevSiblingDim, store: Option<&mut HashMap<&'static str, f32>>, y: &Measurement) -> LayoutResult<f32> {
        let mut reified = 0f32;
        for y in y.iter_adds() {
            reified += y.value.scalar() * match y.unit {
                MeasurementUnit::Units => 1f32,
                MeasurementUnit::Pixels => 1f32 / parent_bounds.column_size.height,
                MeasurementUnit::PercentOfParent => match parent_bounds.bounding_box.height.into_option() {
                    None => Err(LayoutError::new("can't use fraction for y: parent height not known"))?,
                    Some(height) => height / 100f32
                },
                MeasurementUnit::OfPrev => match prev_sibling {
                    PrevSiblingDim::NotApplicable => Err(LayoutError::new("can't use prev for y: not applicable"))?,
                    PrevSiblingDim::FirstChild => 0f32,
                    PrevSiblingDim::Some(prev_sibling_dim) => *prev_sibling_dim
                }
                MeasurementUnit::OfLoad(ident) => match &store {
                    None => Err(LayoutError::new("can't load y: dim-store not applicable"))?,
                    Some(store) => match store.get(ident) {
                        None => Err(LayoutError::new(format!("can't load y: no such dim {}", ident)))?,
                        Some(result) => *result
                    }
                }
            }
        }
        if let Some(ident) = y.store {
            match store {
                None => Err(LayoutError::new("can't store y: dim-store not applicable"))?,
                Some(store) => store.insert(ident, reified)
            };
        }
        Ok(reified)
    }

    fn apply_layout_x(parent_bounds: &ParentBounds, prev_sibling: Option<&Rectangle>, layout: LayoutPosition1D, reified: f32) -> LayoutResult<f32> {
        Ok(match layout {
            LayoutPosition1D::GlobalAbsolute => reified,
            LayoutPosition1D::LocalAbsolute => reified + parent_bounds.bounding_box.x,
            LayoutPosition1D::Relative => match parent_bounds.sub_layout.direction {
                LayoutDirection::Horizontal => match prev_sibling {
                    None => reified + parent_bounds.bounding_box.left().map_err(|err| err.add_dimension("parent.left"))?,
                    Some(prev_sibling) => {
                        // Yes, we do want to reify the parent's sub-layout with it's own bounds
                        let gap = Self::reify_x(parent_bounds, &PrevSiblingDim::NotApplicable, None, &parent_bounds.sub_layout.gap).map_err(|err| err.add_dimension("parent.gap"))?;
                        reified + prev_sibling.right + gap
                    }
                },
                LayoutDirection::Vertical => reified + parent_bounds.bounding_box.x,
                LayoutDirection::Overlap => reified + parent_bounds.bounding_box.x
            }
        })
    }

    fn apply_layout_y(parent_bounds: &ParentBounds, prev_sibling: Option<&Rectangle>, layout: LayoutPosition1D, reified: f32) -> LayoutResult<f32> {
        Ok(match layout {
            LayoutPosition1D::GlobalAbsolute => reified,
            LayoutPosition1D::LocalAbsolute => reified + parent_bounds.bounding_box.y,
            LayoutPosition1D::Relative => match parent_bounds.sub_layout.direction {
                LayoutDirection::Horizontal => reified + parent_bounds.bounding_box.y,
                LayoutDirection::Vertical => match prev_sibling {
                    None => reified + parent_bounds.bounding_box.top().map_err(|err| err.add_dimension("parent.top"))?,
                    Some(prev_sibling) => {
                        // Yes, we do want to reify the parent's sub-layout with it's own bounds
                        let gap = Self::reify_y(parent_bounds, &PrevSiblingDim::NotApplicable, None, &parent_bounds.sub_layout.gap).map_err(|err| err.add_dimension("parent.gap"))?;
                        reified + prev_sibling.bottom + gap
                    }
                },
                LayoutDirection::Overlap => reified + parent_bounds.bounding_box.y
            }
        })
    }
}

impl LayoutPosition {
    pub fn xy(xy: LayoutPosition1D) -> LayoutPosition {
        return LayoutPosition {
            x: xy,
            y: xy
        }
    }
}

impl Default for Bounds {
    fn default() -> Self {
        Bounds {
            layout: LayoutPosition::xy(LayoutPosition1D::Relative),
            x: Measurement::ZERO,
            y: Measurement::ZERO,
            z: 0,
            anchor_x: 0f32,
            anchor_y: 0f32,
            width: None,
            height: None
        }
    }
}


impl Default for LayoutPosition1D {
    fn default() -> Self {
        LayoutPosition1D::Relative
    }
}